store/postgres/src/relational.rs

```diff
@@ -1020,22 +1020,40 @@ impl Layout {
         last_rollup: Option<BlockTime>,
         block_times: &[(BlockNumber, BlockTime)],
     ) -> Result<(), StoreError> {
+        if block_times.is_empty() {
+            return Ok(());
+        }
+
+        // If we have never done a rollup, we can just use the smallest
+        // block time we are getting as the time for the last rollup
+        let mut last_rollup = last_rollup.unwrap_or_else(|| {
+            block_times
+                .iter()
+                .map(|(_, block_time)| *block_time)
+                .min()
+                .unwrap()
+        });
         // The for loop could be eliminated if the rollup queries could deal
         // with the full `block_times` vector, but the SQL for that will be
         // very complicated and is left for a future improvement.
-        let mut last_rollup = last_rollup.unwrap_or(BlockTime::NONE);
         for (block, block_time) in block_times {
             for rollup in &self.rollups {
                 let buckets = rollup.interval.buckets(last_rollup, *block_time);
-                if buckets.is_empty() {
-                    // The rollups are in increasing order of interval size, so
-                    // if a smaller interval doesn't have a bucket between
-                    // last_rollup and block_time, a larger one can't either and
-                    // we are done with this rollup.
-                    break;
-                }
-                for bucket in buckets {
-                    rollup.insert(conn, &bucket, *block)?;
+                // We only need to pay attention to the first bucket; if
+                // there are more buckets, there's nothing to rollup for
+                // them as the next changes we wrote are for `block_time`,
+                // and we'll catch that on the next iteration of the loop.
```

Comment:
My thinking here is this: assume we get passed in `block_times = [b1, b2, b3, .. ]` but `b1` and `b2` are far apart. We call `rollup.interval.buckets(b1, b2)` at some point in the iteration which produces timestamps `[t1, t2, ..]`. Since `b1` and `b2` are far apart, we have something like `t1 <= b1 < t2 < t3 < t4 < t5 <= b2` but we know that there are no writes between `b1` and `b2` - if there were, we'd have some block time in `block_times` between `b1` and `b2`. So we only need to do a rollup for `t1 < b1 < t2`. After that, we set `last_rollup = b2` and repeat the loop for that, which will roll up the bucket `t5 <= b2 < t6`. So there's no need to worry about the buckets starting at `t2`, `t3`, and `t4`.

Maybe there's an error in my logic, but I _think_ that's correct (and tests haven't shown any missed rollups)

Reply:
If you find the above convincing, I should probably add that to the comment

Reply:
Yes its convincing, thanks for the clarification.